lib/elixir/lib/module/types/expr.ex

```diff
@@ -139,27 +139,48 @@ defmodule Module.Types.Expr do
   end
 
   # %{map | ...}
-  def of_expr({:%{}, _, [{:|, _, [_map, args]}]} = expr, stack, context) do
+  def of_expr({:%{}, _, [{:|, _, [map, args]}]} = expr, stack, context) do
     stack = push_expr_stack(expr, stack)
 
-    case of_pairs(args, stack, context) do
-      {:ok, _pairs, context} -> {:ok, {:map, [{:optional, :dynamic, :dynamic}]}, context}
-      {:error, reason} -> {:error, reason}
+    with {:ok, map_type, context} <- of_expr(map, stack, context),
+         {:ok, arg_pairs, context} <- of_pairs(args, stack, context),
+         arg_pairs = pairs_to_unions(arg_pairs, context),
+         dynamic_value_pairs =
+           Enum.map(arg_pairs, fn {:required, key, _value} -> {:required, key, :dynamic} end),
+         args_type = {:map, dynamic_value_pairs ++ [{:optional, :dynamic, :dynamic}]},
+         {:ok, type, context} <- unify(args_type, map_type, stack, context) do
+      {:map, pairs} = resolve_var(type, context)
```

Comment:
This can return a union if we do: `map.foo or map.bar` in a guard and then we update the map. So we need to handle this case too.

Reply:
It can't return a union yet afaict, not until unions are lifted unification. It's hard to tell how to handle this scenario correctly until they unions are lifted. Did you have anything in mind?